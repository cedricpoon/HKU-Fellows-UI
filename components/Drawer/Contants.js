export const LOGO_SIZE = 80;
