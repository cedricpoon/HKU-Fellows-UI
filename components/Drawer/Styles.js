import { StyleSheet } from 'react-native';

import { fontSize } from 'hkufui/theme/grid';
import { logo } from 'hkufui/theme/palette';

export default StyleSheet.create({
  label: {
    fontWeight: 'normal',
    fontSize: fontSize.titleSize
  },
  color: {
    color: logo.green
  },
  container: {
    flex: 1
  },
  scrollBar: {
    flex: 1
  }
});
