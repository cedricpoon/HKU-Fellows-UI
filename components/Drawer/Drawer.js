import React, { PureComponent } from 'react';
import { SafeAreaView, View } from 'react-native';
import { Button, Text, Icon } from 'native-base';
import { DrawerItems } from 'react-navigation';
import PropTypes from 'prop-types';

import NavigationService from 'hkufui/src/NavigationService';
import Logo from '../Logo/Logo';
import Backdrop from '../Backdrop/Backdrop';

import styles from './Styles';
import { LOGO_SIZE } from './Contants';
import { mapLayoutToState } from 'hkufui/components/helper';
import { localize } from 'hkufui/locale';
import { hotPostMinIndex as hot } from 'hkufui/config';

const locale = localize({ language: 'en', country: 'hk' });

export class Drawer extends PureComponent {
  constructor(props) {
    super(props);

    this.state = { drawerLayout: { width: 0, height: 0 } };
  }

  _onLogout = async () => {
    const { onLogout } = this.props;
    try {
      await onLogout();
      // navigate to login screen
      NavigationService.reset('Login');
    } catch (e) { /* ignored */ }
  }

  render() {
    const { username, temperature, token7digits, onTemperature } = this.props;
    const { drawerLayout } = this.state;
    // following react-navigation docs
    return (
      <SafeAreaView
        forceInset={{ top: 'always', horizontal: 'never' }}
        onLayout={mapLayoutToState("drawerLayout", this)}
        style={styles.container}
      >
        <View style={styles.wrapper}>
          <View style={styles.items}>
            <Logo size={LOGO_SIZE} />
            <View style={[styles.userInfoGroup, { height: LOGO_SIZE }]}>
              <Text style={styles.secondaryUserInfo}>{locale['drawer.userInfoIndicator']}</Text>
              <Text style={styles.userInfo}>{username}</Text>
              <Text style={[styles.userInfo, temperature && temperature > hot && styles.hot]}>
                <Icon style={[styles.userInfo, temperature && temperature > hot && styles.hot]} name='ios-flame'/> {temperature === null ? '-' : temperature}
              </Text>
              <Text style={styles.secondaryUserInfo}>
                <Icon style={styles.secondaryUserInfo} name='key' type='Foundation'/> {token7digits}
              </Text>
            </View>
          </View>
          <DrawerItems
            labelStyle={styles.label}
            activeTintColor={styles.color.color}
            {...this.props}
          />
          <Button full transparent dark onPress={onTemperature}>
            <Text style={styles.label}>
              Refresh Temperature
            </Text>
          </Button>
          <Button full transparent dark onPress={this._onLogout}>
            <Text style={styles.label}>
              {locale['drawer.logout']}
            </Text>
          </Button>
        </View>
        <Backdrop
          width={drawerLayout.width}
          height={drawerLayout.height}
          style={styles.backdrop}
        />
      </SafeAreaView>
    );
  }
}

Drawer.defaultProps = {
  username: '-',
  token7digits: '0000000',
  temperature: null
}

Drawer.propTypes = {
  onLogout: PropTypes.func.isRequired,
  onTemperature: PropTypes.func.isRequired,
  username: PropTypes.string,
  token7digits: PropTypes.string,
  temperature: PropTypes.number
}

export default Drawer;
