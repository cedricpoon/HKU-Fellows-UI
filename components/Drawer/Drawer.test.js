import React from 'react';
import { shallow } from 'enzyme';

import Drawer from './Drawer';

describe('Testing Drawer Component', () => {
  const requiredProps = {
    onLogout: () => {},
    onTemperature: () => {}
  };

  it('renders as expected', () => {
    const wrapper = shallow(
      <Drawer {...requiredProps} />
    );
    expect(wrapper).toMatchSnapshot();
  });
});
