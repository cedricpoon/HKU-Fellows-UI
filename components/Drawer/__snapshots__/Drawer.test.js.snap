// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Drawer Component renders as expected 1`] = `
<SafeAreaView
  forceInset={
    Object {
      "horizontal": "never",
      "top": "always",
    }
  }
  onLayout={[Function]}
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "zIndex": 1,
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "flex-start",
          "flexDirection": "row",
          "justifyContent": "space-between",
          "margin": 20,
        }
      }
    >
      <Logo
        size={80}
      />
      <View
        style={
          Array [
            Object {
              "alignItems": "flex-end",
              "flexDirection": "column",
              "justifyContent": "space-between",
            },
            Object {
              "height": 80,
            },
          ]
        }
      >
        <Styled(Text)
          style={
            Object {
              "color": "grey",
              "fontSize": 10,
            }
          }
        >
          You are logged in as
        </Styled(Text)>
        <Styled(Text)
          style={
            Object {
              "fontSize": 12.5,
            }
          }
        >
          -
        </Styled(Text)>
        <Styled(Text)
          style={
            Array [
              Object {
                "fontSize": 12.5,
              },
              null,
            ]
          }
        >
          <Styled(Icon)
            name="ios-flame"
            style={
              Array [
                Object {
                  "fontSize": 12.5,
                },
                null,
              ]
            }
          />
          -
        </Styled(Text)>
        <Styled(Text)
          style={
            Object {
              "color": "grey",
              "fontSize": 10,
            }
          }
        >
          <Styled(Icon)
            name="key"
            style={
              Object {
                "color": "grey",
                "fontSize": 10,
              }
            }
            type="Foundation"
          />
          0000000
        </Styled(Text)>
      </View>
    </View>
    <DrawerNavigatorItems
      activeBackgroundColor="rgba(0, 0, 0, .04)"
      activeTintColor="#3A936B"
      inactiveBackgroundColor="transparent"
      inactiveTintColor="rgba(0, 0, 0, .87)"
      labelStyle={
        Object {
          "fontSize": 15,
          "fontWeight": "normal",
        }
      }
      onLogout={[Function]}
      onTemperature={[Function]}
      temperature={null}
      token7digits="0000000"
      username="-"
    />
    <Styled(Button)
      dark={true}
      full={true}
      onPress={[Function]}
      transparent={true}
    >
      <Styled(Text)
        style={
          Object {
            "fontSize": 15,
            "fontWeight": "normal",
          }
        }
      >
        Refresh Rating
      </Styled(Text)>
    </Styled(Button)>
    <Styled(Button)
      dark={true}
      full={true}
      onPress={[Function]}
      transparent={true}
    >
      <Styled(Text)
        style={
          Object {
            "fontSize": 15,
            "fontWeight": "normal",
          }
        }
      >
        Log out
      </Styled(Text)>
    </Styled(Button)>
  </View>
  <Backdrop
    height={0}
    style={
      Object {
        "zIndex": 0,
      }
    }
    width={0}
  />
</SafeAreaView>
`;
